### PR TITLE
TD-858 : Implement DLS course certificates in DLSV2

### PR DESCRIPTION
**Pull request by Swapnamol**
### JIRA link
https://hee-tis.atlassian.net/browse/TD-858

### Description
1. Create a new view in the refactored platform to display the course certificate with a “Download” link button

2. On “Download”, send the view to the Learning Hub PDF generation service (hiding the download button), await generation and provide as client browser download when ready

### Screenshots
Before change: When we are clicking the "Certificate" or "Preview Certificate" button from the DLS application, it is redirecting to the legacy application.
image
After change:  Created a view to display the certificate and download the certificate. 
image
image
-----
### Developer checks
(Leave tasks unticked if they haven't been appropriate for your ticket.)

I have:
- [x] Run the formatter and made sure there are no IDE errors (see [info on Text Editor settings](https://hee-tis.atlassian.net/wiki/spaces/TP/pages/3546185813/DLS+Dev+Process) to avoid whitespace changes)
- [x] Written tests for the changes (accessibility tests, unit tests for controller, data services, services, view models, etc)
- [x] Manually tested my work with and without JavaScript
- [x] Tested any Views or partials created or changed with [Wave Chrome plugin](https://chrome.google.com/webstore/detail/wave-evaluation-tool/jbbplnpkjmmeebjpijfedlgcdilocofh/related) and addressed any valid accessibility issues
- [ ] Updated/added documentation in [Confluence](https://hee-tis.atlassian.net/wiki/spaces/TP/pages/3546939432/DLS+Code) and/or [GitHub Readme](https://github.com/TechnologyEnhancedLearning/DLSV2/blob/master/README.md). List of documentation links added/changed:
  - [doc_1_here](link_1_here)
- [x] Updated my Jira ticket with information about other parts of the system that were touched as part of the MR and have to be sanity tested to ensure nothing’s broken
- [x] Scanned over my pull request in GitHub and addressed any warnings from the GitHub Build and Test checks.

### CODE REVIEW

**Comment by kevwhitt-hee** [comment]
Morning @Swapnamol, This is looking amazing, thank you. The CI workflow is complaining of 2 broken tests in `DigitalLearningSolutions.Web.Tests` - please could you take a look at them and correct?

**Comment by Swapnamol** [comment]
@kevwhitt-hee  I have added/modified the test cases.

File: DigitalLearningSolutions.Web/appsettings.Development.json [hunk 1]
```diff
@@ -19,5 +19,10 @@
     "PricingPage": true,
     "RefactoredFindYourCentrePage": true
   },
-  "LearningHubOpenAPIBaseUrl": "https://uks-learninghubnhsuk-openapi-test.azurewebsites.net"
+  "LearningHubOpenAPIBaseUrl": "https://uks-learninghubnhsuk-openapi-test.azurewebsites.net",
+  "LearningHubReportAPIConfig": {
+    "BaseUrl": "https://reportapi-test.test-learninghub.org.uk",
+    "ClientId": "BCDE972F-E007-4DFF-B233-8870474AEEAA",
+    "ClientIdentityKey": "56B1C9EF-9F25-4F8C-B771-85DCB2097B2E"
```
Comment (kevwhitt-hee) [inline DigitalLearningSolutions.Web/appsettings.Development.json:26]: @Swapnamol, is this a secret? Does it need moving to secrets.config?

**Review by kevwhitt-hee — CHANGES_REQUESTED** [review]
Looks like there's a build error, now on the CI pipeline. I'm getting the same error when I try to build the branch locally: 
image
Please take a look and fix.

File: DigitalLearningSolutions.Web/appsettings.Development.json (hunk 1 above)
Comment (Swapnamol) [reply DigitalLearningSolutions.Web/appsettings.Development.json]: @kevwhitt-hee ClientIdentityKey moved to secrets.config.

**Comment by Swapnamol** [comment]
@kevwhitt-hee - Fixed the CI build issues also.

File: DigitalLearningSolutions.Web/Controllers/Certificate/CertificateController.cs [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+           ICertificateService certificateService
+       )
+        {
+            this.centresDataService = centresDataService;
+            this.mapsApiHelper = mapsApiHelper;
+            this.logger = logger;
+            this.imageResizeService = imageResizeService;
+            this.certificateService = certificateService;
+        }
+        [Route("preview:int")]
```
Comment (kevwhitt-hee) [inline DigitalLearningSolutions.Web/Controllers/Certificate/CertificateController.cs:47]: Could we go for `[Route("Certificate/Preview")]`, please?

File: DigitalLearningSolutions.Web/Controllers/Certificate/CertificateController.cs [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+            var certificateInfo = certificateService.GetPreviewCertificateForCentre(centreId);
+            if (certificateInfo == null)
+            {
+                return NotFound();
+            }
+
+            var model = new PreviewCertificateViewModel(certificateInfo);
+            return View("Index", model);
+        }
+        [Route("progressId:int")]
```
Comment (kevwhitt-hee) [inline DigitalLearningSolutions.Web/Controllers/Certificate/CertificateController.cs:60]: Could we go for [Route("Certificate/Activity")], please?

**Review by kevwhitt-hee — COMMENTED** [review]
Absolutely thrilled to see this working. Four tiny observations:

- When opening a certificate from the "Completed" activities list in the Learning Portal, we should open in a new window, since there is no navigation to return to it in the certificate view.
- It would be nice to sort the URL for the preview and progress pages. They have the route templates `preview:int` and `progress:int` which are URL encoded in the browser and look strange.
- The font styling for the username and course title don't look right in the pdf certificate:
  - Razor view: 
image
   - PDF: 
image
- Finally, to make it fit better with the refactored DLS site, please could we adjust the razor page background colour from `#d9ecff` to `#f0f4f5`

File: DigitalLearningSolutions.Web/Controllers/Certificate/CertificateController.cs (hunk 2 above)
Comment (Swapnamol) [reply DigitalLearningSolutions.Web/Controllers/Certificate/CertificateController.cs]: @kevwhitt-hee - Implemented.

File: DigitalLearningSolutions.Web/Controllers/Certificate/CertificateController.cs (hunk 3 above)
Comment (Swapnamol) [reply DigitalLearningSolutions.Web/Controllers/Certificate/CertificateController.cs]: @kevwhitt-hee - Implemented

**Comment by Swapnamol** [comment]
@kevwhitt-hee - implemented all observations.

File: DigitalLearningSolutions.Web/appsettings.json [hunk 4]
```diff
@@ -30,5 +30,10 @@
     "ClientCode": "DigitalLearningSolutions"
   },
   "ExcelPassword": "0f8cc6f0-9f21-4f0f-9cb9-365675490458",
-  "IsTransactionScope": true
+  "IsTransactionScope": true,
+  "LearningHubReportAPIConfig": {
+    "BaseUrl": "https://reportapi-test.test-learninghub.org.uk",
```
Comment (kevwhitt-hee) [inline DigitalLearningSolutions.Web/appsettings.json:35]: Could we replace this with the live equivalent to reduce the config changes required during deployment?

Comment (Swapnamol) [reply DigitalLearningSolutions.Web/appsettings.json]: Replaced with the production URL @kevwhitt-hee

**Review by kevwhitt-hee — APPROVED** [review]

**Pull request merged**